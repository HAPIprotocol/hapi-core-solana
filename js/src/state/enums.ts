export enum HapiAccountType {
  Uninitialized,
  Community,
  Network,
  Reporter,
  Case,
  Address,
}

export enum ReporterType {
  Inactive,
  Tracer,
  Full,
  Authority,
}

export enum Category {
  // Tier 0
  /// Safe
  Safe,

  // Tier 1 - Low risk
  /// Wallet service - custodial or mixed wallets
  WalletService,

  /// Merchant service
  MerchantService,

  /// Mining pool
  MiningPool,

  /// Exchange (Low Risk) - Exchange with high KYC standards
  LowRiskExchange,

  // Tier 2 - Medium risk
  /// Exchange (Medium Risk)
  MediumRiskExchange,

  /// DeFi application
  DeFi,

  /// OTC Broker
  OTCBroker,

  /// Cryptocurrency ATM
  ATM,

  /// Gambling
  Gambling,

  // Tier 3 - High risk
  /// Illicit organization
  IllicitOrganization,

  /// Mixer
  Mixer,

  /// Darknet market or service
  DarknetService,

  /// Scam
  Scam,

  /// Ransomware
  Ransomware,

  /// Theft - stolen funds
  Theft,

  // Tier 4 - Severe risk
  /// Terrorist financing
  TerroristFinancing,

  /// Sanctions
  Sanctions,

  /// Child abuse and porn materials
  ChildAbuse,
}
