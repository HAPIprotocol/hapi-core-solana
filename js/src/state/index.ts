export * from "./address";
export * from "./case";
export * from "./community";
export * from "./enums";
export * from "./network";
export * from "./reporter";
