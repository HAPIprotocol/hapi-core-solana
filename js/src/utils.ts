import BN from "bn.js";
import { decode } from "bs58";
import {
  Connection,
  Keypair,
  PublicKey,
  Transaction,
  TransactionInstruction,
} from "@solana/web3.js";

export class u8 extends BN {
  static size = 1;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u8.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u8.size);
  }

  /**
   * Construct a Numberu8 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u8.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(4)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u32 extends BN {
  static size = 4;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u32.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u32.size);
  }

  /**
   * Construct a Numberu32 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u32.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u64 extends BN {
  static size = 8;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u64.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u64.size);
  }

  /**
   * Construct a Numberu64 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u64.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export function bnToBuffer(bn: BN, size: number): Buffer {
  const a = bn.toArray().reverse();
  const b = Buffer.from(a);
  if (b.length === size) {
    return b;
  }
  if (b.length > size) {
    throw new Error(`Buffer too large (size: ${size})`);
  }

  const zeroPad = Buffer.alloc(size);
  b.copy(zeroPad);
  return zeroPad;
}

export const signAndSendTransactionInstructions = async (
  // sign and send transaction
  connection: Connection,
  signers: Array<Keypair>,
  feePayer: Keypair,
  txInstructions: Array<TransactionInstruction>
): Promise<string> => {
  const tx = new Transaction();
  tx.feePayer = feePayer.publicKey;
  signers.push(feePayer);
  tx.add(...txInstructions);
  return await connection.sendTransaction(tx, signers, {
    preflightCommitment: "single",
  });
};

export function mapToBuffer<K extends number, V extends number | boolean>(
  map: Map<K, V>
): Buffer {
  const buffer = Buffer.alloc(4 + map.size * 2);

  // Write size header
  buffer.writeUInt32LE(map.size);

  // Write map KVs
  let i = 4;
  const keys = Array.from(map.keys()).sort((a, b) => a - b);
  for (const key of keys) {
    buffer.writeUInt8(key, i);
    buffer.writeUInt8(map.get(key) ? 1 : 0, i + 1);
    i += 2;
  }

  return buffer;
}

export function setToBuffer<T extends number>(set: Set<T>, size = 4): Buffer {
  let value = new BN(0);

  for (const i of set.values()) {
    value = value.or(new BN(i));
  }

  return value.toBuffer("le", size);
}

export function base58ToPublicKey(address: string): PublicKey {
  const buffer = decode(address);
  return new PublicKey(buffer);
}

export function hexToPublicKey(hex: string): PublicKey {
  const byteLength = Buffer.from(hex, "hex").length;
  if (byteLength > 32) {
    throw RangeError(
      `hexToPublicKey: input length can't be longer than 32 bytes`
    );
  }
  const buffer = Buffer.alloc(32);
  buffer.write(hex, 32 - byteLength, "hex");
  return PublicKey.decodeUnchecked(buffer);
}
