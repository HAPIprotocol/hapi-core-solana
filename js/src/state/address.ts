import { Numberu64, Numberu8 } from "../utils";
import { Category, HapiAccountType } from "./enums";

export class Address {
  /// HAPI account type
  account_type: HapiAccountType;

  /// Risk score
  risk: Numberu8;

  /// Case ID
  case_id: Numberu64;

  /// Category
  category: Category;
}
