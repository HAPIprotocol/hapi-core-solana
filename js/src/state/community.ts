import { Connection, PublicKey } from "@solana/web3.js";
import { serialize, deserializeUnchecked } from "borsh";
import BN from "bn.js";

import { u64 } from "../utils";
import { HapiAccountType } from "./enums";

class CommunityState {
  account_type: number;
  authority: Uint8Array;
  name: string;
  next_case_id: BN;
  constructor(object: Partial<CommunityState>) {
    Object.assign(this, object);
  }
  static schema = new Map([
    [
      CommunityState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["authority", [32]],
          ["next_case_id", "u64"],
          ["name", "string"],
        ],
      },
    ],
  ]);
  static size = 73;
}

export class Community {
  /// HAPI account type
  accountType = HapiAccountType.Community;

  /// HAPI authority account
  authority: PublicKey;

  /// ID of the next created case
  nextCaseId: u64;

  /// HAPI community name
  name: string;

  static size = CommunityState.size;

  constructor(data?: Partial<Community>) {
    if (data) {
      this.authority = data.authority;
      this.nextCaseId = data.nextCaseId;
      this.name = data.name;
    }
  }

  static async getAddress(
    programId: PublicKey,
    communityName: string
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      Community.getAddressSeeds(communityName),
      programId
    );
  }

  static getAddressSeeds(communityName: string): Uint8Array[] {
    return [Buffer.from("community"), Buffer.from(communityName)];
  }

  static fromState(state: CommunityState): Community {
    return new Community({
      authority: new PublicKey(state.authority),
      nextCaseId: new u64(state.next_case_id),
      name: state.name,
    });
  }

  static deserialize(buffer: Buffer): Community {
    return Community.fromState(
      deserializeUnchecked(CommunityState.schema, CommunityState, buffer)
    );
  }

  static async retrieve(
    programId: PublicKey,
    connection: Connection,
    communityName: string
  ): Promise<{ data: Community; account: PublicKey }> {
    const [communityAddress] = await Community.getAddress(
      programId,
      communityName
    );

    const communityAccount = await connection.getAccountInfo(communityAddress);
    if (!communityAccount) {
      throw new Error(
        `Community not found: ${communityName} (${communityAddress})`
      );
    }

    return {
      data: Community.deserialize(communityAccount.data),
      account: communityAddress,
    };
  }

  serialize(): Uint8Array {
    const buf = Buffer.alloc(CommunityState.size);
    buf.set(serialize(CommunityState.schema, this.toState()));
    return buf;
  }

  toState(): CommunityState {
    return new CommunityState({
      account_type: this.accountType,
      authority: this.authority.toBytes(),
      next_case_id: this.nextCaseId,
      name: this.name,
    });
  }
}
