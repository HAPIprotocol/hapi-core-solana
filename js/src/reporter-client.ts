import { PublicKey } from "@solana/web3.js";

import { HapiActionResponse, ReaderClient } from "./reader-client";
import { Case, Category } from "./state";

/** HAPI client to operate reporter program functions on Solana */
export class ReporterClient extends ReaderClient {
  async createCase(
    name: string,
    categories: Category[],
    communityName?: string
  ): Promise<HapiActionResponse<Case>> {
    name;
    categories;
    communityName;
    // TODO: create an instruction
    return {
      account: new PublicKey("0"),
    };
  }

  async updateCase(): Promise<void> {
    // TODO: create an instruction
  }

  async createAddress(): Promise<void> {
    // TODO: create an instruction
  }

  async updateAddress(): Promise<void> {
    // TODO: create an instruction
  }
}
