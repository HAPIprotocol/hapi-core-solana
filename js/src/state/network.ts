import { Connection, PublicKey } from "@solana/web3.js";
import { deserializeUnchecked, Schema, serialize } from "borsh";

import { Community } from "./community";
import { HapiAccountType } from "./enums";

class NetworkState {
  account_type: number;
  name: string;
  constructor(object: Partial<NetworkState>) {
    Object.assign(this, object);
  }
  static schema: Schema = new Map([
    [
      NetworkState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["name", "string"],
        ],
      },
    ],
  ]);
  static size = 33;
}

export class Network {
  /// HAPI account type
  accountType = HapiAccountType.Network;

  /// HAPI network name
  name: string;

  static size = NetworkState.size;

  constructor(object?: Partial<Network>) {
    if (object) {
      Object.assign(this, object);
    }
  }

  static async getAddress(
    programId: PublicKey,
    communityAddress: PublicKey,
    networkName: string
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from("network"),
        communityAddress.toBuffer(),
        Buffer.from(networkName),
      ],
      programId
    );
  }

  static fromState(state: NetworkState): Network {
    const network = new Network();
    network.accountType = state.account_type;
    network.name = state.name;
    return network;
  }

  static deserialize(buffer: Buffer): Network {
    return Network.fromState(
      deserializeUnchecked(NetworkState.schema, NetworkState, buffer)
    );
  }

  static async retrieve(
    programId: PublicKey,
    connection: Connection,
    communityName: string,
    networkName: string
  ): Promise<{ data: Network; account: PublicKey }> {
    const [communityAddress] = await Community.getAddress(
      programId,
      communityName
    );

    const [networkAddress] = await Network.getAddress(
      programId,
      communityAddress,
      networkName
    );

    const account = await connection.getAccountInfo(networkAddress);
    if (!account) {
      throw new Error(
        `Network not found: "${networkName}" (${networkAddress}) in community "${communityName}" (${communityAddress})`
      );
    }

    return { data: Network.deserialize(account.data), account: networkAddress };
  }

  serialize(): Uint8Array {
    const buf = Buffer.alloc(NetworkState.size);
    buf.set(serialize(NetworkState.schema, this.toState()));
    return buf;
  }

  toState(): NetworkState {
    return new NetworkState({
      account_type: this.accountType,
      name: this.name,
    });
  }
}
