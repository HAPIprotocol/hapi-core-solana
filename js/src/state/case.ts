import { Connection, PublicKey } from "@solana/web3.js";
import { deserialize, serialize } from "borsh";

import { HAPI_PROGRAM_ID } from "../constants";
import { mapToBuffer, u64 } from "../utils";
import { Community } from "./community";
import { Categories, Category, HapiAccountType } from "./enums";
import { Network } from "./network";

class CaseState {
  account_type: number;
  name: string;
  reporter_key: Uint8Array;
  categories: Uint8Array;
  constructor(object: Partial<CaseState>) {
    Object.assign(this, object);
  }
  static schema = new Map([
    [
      CaseState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["name", "string"],
          ["reporter_key", [32]],
          ["categories", "u32"],
        ],
      },
    ],
  ]);
}

export class Case {
  /// HAPI account type
  accountType: HapiAccountType;

  /// Case name
  name: string;

  /// Case reporter key
  reporterKey: PublicKey;

  /// Categories
  categories: Map<Category, boolean>;

  constructor(data?: Partial<Case>) {
    if (data) {
      Object.assign(this, data);
      for (const category of Categories) {
        if (!this.categories.has(category)) {
          this.categories.set(category, false);
        }
      }
    }
  }

  static async getAddress(
    networkAddress: PublicKey,
    caseId: u64
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [Buffer.from("case"), networkAddress.toBuffer(), caseId.toBuffer()],
      HAPI_PROGRAM_ID
    );
  }

  static fromState(state: CaseState): Case {
    return new Case({
      accountType: state.account_type,
      name: state.name,
      reporterKey: new PublicKey(state.reporter_key),
      categories: new Map<Category, boolean>(),
    });
  }

  static deserialize(buffer: Buffer): Case {
    return Case.fromState(deserialize(CaseState.schema, CaseState, buffer));
  }

  static async retrieve(
    connection: Connection,
    communityName: string,
    networkName: string,
    caseId: u64
  ): Promise<Case> {
    const [communityAddress] = await Community.getAddress(communityName);

    const [networkAddress] = await Network.getAddress(
      communityAddress,
      networkName
    );

    const [caseAddress] = await Case.getAddress(networkAddress, caseId);

    const account = await connection.getAccountInfo(caseAddress, "processed");
    if (!account) {
      throw new Error("Invalid case account provided");
    }

    return Case.deserialize(account.data);
  }

  serialize(): Uint8Array {
    return serialize(CaseState.schema, this.toState());
  }

  toState(): CaseState {
    return new CaseState({
      account_type: this.accountType,
      name: this.name,
      reporter_key: this.reporterKey.toBytes(),
      categories: mapToBuffer(this.categories),
    });
  }
}
