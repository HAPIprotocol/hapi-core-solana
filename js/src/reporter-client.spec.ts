//jest.mock("rpc-websockets"); // disable web3's websocket connections

import stringify from "fast-json-stable-stringify";
import nock from "nock";

import {
  captureConsoleError,
  mockConfirmTransaction,
  mockRpcAccount,
  mockRpcError,
  mockRpcOk,
} from "../test/util/mocks";
import { Case, CaseStatus, Category, Community } from "./state";
import { ReporterClient } from "./reporter-client";
import {
  AUTHORITY,
  REPORTER_ALICE,
  REPORTER_CAROL,
  UNINITIALIZED,
} from "../test/keypairs";
import { u64 } from "./utils";

describe("ReporterClient", () => {
  let client: ReporterClient;

  const endpoint = "http://localhost:8899";

  beforeEach(async () => {
    nock.disableNetConnect();
    client = new ReporterClient({
      endpoint,
      payer: UNINITIALIZED,
      communityName: "hapi.one",
      commitment: "processed",
    });
  });

  afterEach(() => {
    nock.cleanAll();
    jest.clearAllMocks();
  });

  it("should initialize", () => {
    expect(client).toBeDefined();
  });

  describe("createCase", () => {
    it("should throw - community not found", async () => {
      client.switchCommunity("community404");

      mockRpcAccount(
        endpoint,
        "mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh",
        null
      );

      await expect(() =>
        client.createCase("erroneous case", CaseStatus.Open, [])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid name", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(10),
        })
      );

      await expect(() =>
        client.createCase("loooooong looooooong maaaaaaan", CaseStatus.Open, [])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid categories", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(10),
        })
      );

      await expect(() =>
        client.createCase("invalid categories", CaseStatus.Open, [4294967295])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - reporter doesn't have rights", async () => {
      client.payer = REPORTER_CAROL;

      const consoleSpy = captureConsoleError();

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(10),
        })
      );

      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "FMjzqMLtuECN4cKNZcu221QCBVExhJZqrKACi7pmw4qL",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcError(endpoint, "sendTransaction", [], {
        code: -32002,
        message:
          "Transaction simulation failed: Error processing Instruction 0: custom program error: 0x9",
        data: {
          accounts: null,
          err: { InstructionError: [0, { Custom: 9 }] },
          logs: [
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 invoke [1]",
            'Program log: HAPI-INSTRUCTION: CreateCase { case_id: 47, categories: 32, status: Open, name: "invalid reporter" }',
            "Program log: Reporter doesn't have a permission to report a case",
            "Program log: HAPI-ERROR: InvalidReporterPermissions",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 consumed 13764 of 200000 compute units",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 failed: custom program error: 0x9",
          ],
        },
      });

      await expect(() =>
        client.createCase("invalid reporter", CaseStatus.Open, [Category.DeFi])
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(consoleSpy.finish()).toContain(
        "Reporter doesn't have a permission to report a case"
      );
    });

    it("should create a case - success", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(48),
        })
      );

      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "FMjzqMLtuECN4cKNZcu221QCBVExhJZqrKACi7pmw4qL",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcOk(
        endpoint,
        "sendTransaction",
        [],
        "43peZU1kygGZ5WZA4spohczd4iKHG3RuoX1XbNGcuMsghnsRKbE2PD9jfTDDUTXLaucDX7WY4a3ENtkDG7wPDKGU"
      );

      mockRpcAccount(
        endpoint,
        "E3NTxagJJwyQrwFC4e2ndpYR3hWoJC59DrQ3aCabjPxo",
        new Case({
          name: "Slice the dice",
          reporterKey: client.payerPublicKey,
          status: CaseStatus.Open,
          categories: [Category.Gambling],
        })
      );

      mockConfirmTransaction(client);

      mockRpcAccount(
        endpoint,
        "DMYBunz3NcxbuR9yFt8AUCfszJV8VFzm47WWHmisMnhv",
        new Case({
          name: "Slice the dice",
          reporterKey: client.payerPublicKey,
          status: CaseStatus.Open,
          categories: [Category.Gambling],
        })
      );

      const {
        data,
        meta: { caseId },
      } = await client.createCase("Slice the dice", CaseStatus.Open, [
        Category.Gambling,
      ]);

      expect(stringify(data)).toMatchSnapshot();
      expect(caseId.toString()).toEqual("48");
    });
  });

  describe("updateCase", () => {
    it("should throw - community not found", async () => {
      client.switchCommunity("community404");

      mockRpcAccount(
        endpoint,
        "mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh",
        null
      );

      await expect(() =>
        client.updateCase(new u64(1), CaseStatus.Closed, [
          Category.MediumRiskExchange,
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - case not found", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(48),
        })
      );

      mockRpcAccount(
        endpoint,
        "5yGEyZT2WwuLD6BJuvP7TXRcR2ojZTpFVv2TkQm6eX2P",
        null
      );

      await expect(() =>
        client.updateCase(new u64(404), CaseStatus.Closed, [
          Category.MediumRiskExchange,
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid categories", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(10),
        })
      );

      mockRpcAccount(
        endpoint,
        "6vGsVQ1YMu5zkNUMJ5j5H1TVimfennBcYuYP9hXw1kB2",
        new Case({
          name: "super case",
          reporterKey: client.payerPublicKey,
          status: CaseStatus.Open,
          categories: [],
        })
      );

      await expect(() =>
        client.updateCase(new u64(1), CaseStatus.Closed, [4294967295])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid status", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(11),
        })
      );

      mockRpcAccount(
        endpoint,
        "DAgLvQ9zU2K5mseeBjCXubv8bCx41F6zwWbkE2ZeuKY2",
        new Case({
          name: "super case",
          reporterKey: client.payerPublicKey,
          status: CaseStatus.Open,
          categories: [],
        })
      );

      await expect(() =>
        client.updateCase(new u64(9), 420, [])
      ).rejects.toThrowErrorMatchingSnapshot();

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(11),
        })
      );

      mockRpcAccount(
        endpoint,
        "DAgLvQ9zU2K5mseeBjCXubv8bCx41F6zwWbkE2ZeuKY2",
        new Case({
          name: "super case",
          reporterKey: client.payerPublicKey,
          status: CaseStatus.Open,
          categories: [],
        })
      );

      await expect(() =>
        client.updateCase(new u64(9), 3, [])
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should update a case - success", async () => {
      client.payer = REPORTER_ALICE;

      mockRpcAccount(
        endpoint,
        "DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja",
        new Community({
          authority: AUTHORITY.publicKey,
          name: "hapi.one",
          nextCaseId: new u64(11),
        })
      );

      mockRpcAccount(
        endpoint,
        "HqvJ9EVws3tEUuptmXUZpPh3JcJJvoSPBk5LTgaoAtV9",
        new Case({
          name: "gamble",
          status: CaseStatus.Open,
          categories: [],
          reporterKey: client.payerPublicKey,
        })
      );

      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "FMjzqMLtuECN4cKNZcu221QCBVExhJZqrKACi7pmw4qL",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcOk(
        endpoint,
        "sendTransaction",
        [],
        "43peZU1kygGZ5WZA4spohczd4iKHG3RuoX1XbNGcuMsghnsRKbE2PD9jfTDDUTXLaucDX7WY4a3ENtkDG7wPDKGU"
      );

      mockConfirmTransaction(client);

      mockRpcAccount(
        endpoint,
        "HqvJ9EVws3tEUuptmXUZpPh3JcJJvoSPBk5LTgaoAtV9",
        new Case({
          name: "gamble",
          status: CaseStatus.Closed,
          categories: [Category.Gambling],
          reporterKey: client.payerPublicKey,
        })
      );

      const { data } = await client.updateCase(new u64(2), CaseStatus.Closed, [
        Category.Gambling,
      ]);

      expect(stringify(data)).toMatchSnapshot();
    });
  });

  describe("createAddress", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - network not found");
    it.todo("should throw - invalid address");
    it.todo("should throw - invalid category");
    it.todo("should throw - invalid risk score");
    it.todo("should create an address - success");
  });

  describe("updateAddress", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - network not found");
    it.todo("should throw - address not found");
    it.todo("should throw - invalid category");
    it.todo("should throw - invalid risk score");
    it.todo("should update an address - success");
  });
});
