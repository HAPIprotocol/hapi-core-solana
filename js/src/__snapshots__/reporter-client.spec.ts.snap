// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReporterClient createCase should create a case - success 1`] = `"{\\"accountType\\":4,\\"categories\\":[256],\\"name\\":\\"Slice the dice\\",\\"reporterKey\\":{\\"_bn\\":\\"c0fdcf81e4689a79e0cb64d88098f5342910652a340ddc0703dbdb10386dc678\\"},\\"status\\":0}"`;

exports[`ReporterClient createCase should throw - community not found 1`] = `"Community not found: community404 (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReporterClient createCase should throw - invalid categories 1`] = `"Unknown category 4294967295"`;

exports[`ReporterClient createCase should throw - invalid name 1`] = `"Case name length should not be over 28 bytes"`;

exports[`ReporterClient createCase should throw - reporter doesn't have rights 1`] = `"failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x9"`;

exports[`ReporterClient updateCase should throw - case not found 1`] = `"Invalid case account provided"`;

exports[`ReporterClient updateCase should throw - community not found 1`] = `"Community not found: community404 (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReporterClient updateCase should throw - invalid categories 1`] = `"Unknown category 4294967295"`;

exports[`ReporterClient updateCase should throw - invalid status 1`] = `"Unknown case status 420"`;

exports[`ReporterClient updateCase should throw - invalid status 2`] = `"Unknown case status 3"`;

exports[`ReporterClient updateCase should update a case - success 1`] = `"{\\"accountType\\":4,\\"categories\\":[256],\\"name\\":\\"gamble\\",\\"reporterKey\\":{\\"_bn\\":\\"c0fdcf81e4689a79e0cb64d88098f5342910652a340ddc0703dbdb10386dc678\\"},\\"status\\":1}"`;
